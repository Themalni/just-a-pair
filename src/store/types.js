export const SET_USER = 'setUser';
export const CLEAR_USER = 'clearUser';
export const ADD_TO_CART = 'addToCart';
export const REMOVE_FROM_CART = 'removeFromCart';
export const SET_PRODUCT_TO_PREVIEW = 'setProductToPreview';