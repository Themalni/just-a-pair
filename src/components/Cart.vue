<template>
  <div>
    <!-- Modal -->
    <div
      class="modal fade"
      id="shoppingCartModal"
      tabindex="-1"
      role="dialog"
      aria-labelledby="exampleModalLabel"
      aria-hidden="false"
    >
      <div class="modal-dialog modal-frame modal-top" role="document">
        <div class="modal-content">
          <div class="modal-body">
            <table class="table">
              <thead class="main-blue white-text">
                <tr>
                  <th scope="col" class="narrow-col">Image</th>
                  <th scope="col" class="wide-col">Title</th>
                  <th scope="col" class="narrow-col">Color</th>
                  <th scope="col" class="narrow-col">Size</th>
                  <th scope="col" class="narrow-col">Price</th>
                  <th scope="col" class="narrow-col">Remove</th>
                </tr>
              </thead>
              <tbody name="fade">
                <tr v-for="(item, index) in cart" v-bind="cart" :key="index">
                  <th scope="row">
                    <img :src="item.image" class="cart-item-img">
                  </th>
                  <td>{{ item.title }}</td>
                  <td>{{ item.color }}</td>
                  <td>{{ item.selectedSize }}</td>
                  <td>${{ item.price }}</td>
                  <td scope="col">
                    <button
                      class="btn light-blue btn-sm btn-remove"
                      @click="removeFromCart(index)"
                    >X</button>
                  </td>
                </tr>
              </tbody>
              <tfoot>
                <tr v-show="total > 0">
                  <th id="total" colspan="4" class="text-right">Total:</th>
                  <td colspan="2">${{ total }}</td>
                </tr>
                <tr v-show="total === 0">
                  <th colspan="5" class="text-center">Your shopping cart is emty</th>
                </tr>
              </tfoot>
            </table>
          </div>
          <div class="modal-footer">
            <button
              type="button"
              class="btn orange-accent btn-sm"
              data-dismiss="modal"
            >Continue Shopping</button>
            <button type="button" class="btn main-blue btn-sm">Checkout</button>
          </div>
        </div>
      </div>
    </div>
    <!-- /Modal -->
  </div>
</template>

<script>
import { mapGetters } from "vuex";

export default {
  name: "Cart",
  computed: {
    ...mapGetters(["products", "inCart"]),
    cart() {
      // return this.$store.getters.inCart.map(cartItem => {
      //   return this.$store.getters.products.find(itemToAdd => {
      //     return cartItem === itemToAdd.id;
      //   });
      // });
      return this.inCart;
    },
    total() {
      return this.cart.reduce((acc, cur) => acc + cur.price, 0);
    }
  },
  methods: {
    removeFromCart(index) {
      this.$store.dispatch("removeFromCart", index);
    }
  }
};
</script>

<style lang="scss" scoped>
.modal-body {
  @include device-size(xs) {
    padding: 0;
  }
  @include device-size(xs, sm) {
    overflow: scroll;
  }
}
.modal-dialog {
  @include device-size(xs) {
    margin: 0;
  }
}
.btn {
  &.btn-remove {
    padding: 0.55em 1em;
  }
}
.cart-item-img {
  width: 80px;
  height: 100%;
}
.table {
  transition: 0.3s linear;
  overflow: auto;

  @include device-size(lg, xl) {
    table-layout: fixed;
    width: 100%;
  }

  td {
    vertical-align: middle;
    font-weight: bold;
    font-size: 1.1em;
    color: #444;

    @include device-size(xs) {
      font-size: 1em;
    }
  }
  th {
    @include device-size(lg, xl) {
      min-width: 200px;
    }
  }
  tfoot {
    th,
    td {
      vertical-align: middle;
      font-weight: bold;
      font-size: 1.4em;
      color: #444;
    }
  }
  .wide-col {
    @include device-size(lg, xl) {
      width: 40%;
    }
  }
  .narrow-col {
    @include device-size(lg, xl) {
      width: 10%;
    }
  }
}
.modal-content {
  transition: 0.3s linear;
}
</style>