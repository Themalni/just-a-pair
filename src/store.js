import Vue from 'vue';
import Vuex from 'vuex';
import VuexPersist from 'vuex-persist'

Vue.use(Vuex);

const vuexLocalStorage = new VuexPersist({ storage: localStorage })

export const store = new Vuex.Store({
    state: {
        products: [{
            id: 621,
            offer: "Sale",
            title: "Be Yourself",
            sku: "AK55-36-Blue",
            stock: "in stock",
            price: 89,
            salePercentage: 65,
            color: "Blue",
            image: '../assets/sneakers.jpg',
            sliderImages: [
                '../assets/sneakers.jpg',
                '../assets/sneakers.jpg',
                '../assets/sneakers.jpg'
            ],
            sale: true,
            productRating: "4",
            selectedSize: '',
            sizes: [{
                number: '36',
                available: true
            },
            {
                number: '37',
                available: true
            },
            {
                number: '38',
                available: true
            },
            {
                number: '39',
                available: true
            },
            {
                number: '40',
                available: true
            },
            {
                number: '41',
                available: true
            },
            {
                number: '42',
                available: false
            }
            ]
        },
        {
            id: 622,
            offer: "New",
            title: "Star Gazers",
            sku: "AL33-36-White",
            stock: "in stock",
            price: 139,
            salePercentage: '',
            color: "White",
            image: '../assets/shoes_white_1.jpg',
            sliderImages: [
                '../assets/shoes_white_1.jpg',
                '../assets/shoes_white_1.jpg',
                '../assets/shoes_white_1.jpg'
            ],
            sale: false,
            productRating: "5",
            selectedSize: '',
            sizes: [{
                number: '36',
                available: true
            },
            {
                number: '37',
                available: true
            },
            {
                number: '38',
                available: true
            },
            {
                number: '39',
                available: true
            },
            {
                number: '40',
                available: false
            },
            {
                number: '41',
                available: true
            },
            {
                number: '42',
                available: false
            }
            ]
        },
        {
            id: 623,
            offer: "Sale",
            title: "Millenium Night",
            sku: "AJ11-36-Red",
            stock: "in stock",
            price: 99,
            salePercentage: 15,
            color: "Red",
            image: '../assets/shoes_red_1.jpg',
            sliderImages: [
                '../assets/shoes_red_1.jpg',
                '../assets/shoes_red_1.jpg',
                '../assets/shoes_red_1.jpg'
            ],
            sale: true,
            productRating: "5",
            selectedSize: '',
            sizes: [{
                number: '36',
                available: true
            },
            {
                number: '37',
                available: false
            },
            {
                number: '38',
                available: true
            },
            {
                number: '39',
                available: true
            },
            {
                number: '40',
                available: true
            },
            {
                number: '41',
                available: true
            },
            {
                number: '42',
                available: false
            }
            ]
        },
        {
            id: 624,
            offer: "New",
            title: "Be Diva Stillettos",
            sku: "AJ11-36-Blue",
            stock: "out of stock",
            price: 84,
            salePercentage: '',
            color: "Blue",
            image: '../assets/shoes_blue_1.jpg',
            sliderImages: [
                '../assets/shoes_blue_1.jpg',
                '../assets/shoes_blue_1.jpg',
                '../assets/shoes_blue_1.jpg'
            ],
            sale: false,
            productRating: "4",
            selectedSize: '',
            sizes: [{
                number: '36',
                available: true
            },
            {
                number: '37',
                available: true
            },
            {
                number: '38',
                available: true
            },
            {
                number: '39',
                available: false
            },
            {
                number: '40',
                available: true
            },
            {
                number: '41',
                available: true
            },
            {
                number: '42',
                available: false
            }
            ]
        },
        {
            id: 625,
            offer: "Sale",
            title: "Solarine",
            sku: "AG23-36-Black",
            stock: "out of stock",
            price: 249,
            salePercentage: 70,
            color: "Black",
            image: '../assets/shoes_black_1.jpg',
            sliderImages: [
                '../assets/shoes_black_1.jpg',
                '../assets/shoes_black_1.jpg',
                '../assets/shoes_black_1.jpg'
            ],
            sale: true,
            productRating: "4",
            selectedSize: '',
            sizes: [{
                number: '36',
                available: true
            },
            {
                number: '37',
                available: true
            },
            {
                number: '38',
                available: false
            },
            {
                number: '39',
                available: true
            },
            {
                number: '40',
                available: true
            },
            {
                number: '41',
                available: false
            },
            {
                number: '42',
                available: false
            }
            ]
        },
        {
            id: 626,
            offer: "New",
            title: "Symphony Light",
            sku: "AB75-36-Grey",
            stock: "in stock",
            price: 129,
            salePercentage: '',
            color: "Grey",
            image: '../assets/shoes_grey_1.jpg',
            sliderImages: [
                '../assets/shoes_grey_1.jpg',
                '../assets/shoes_grey_1.jpg',
                '../assets/shoes_grey_1.jpg'
            ],
            sale: false,
            productRating: "5",
            selectedSize: '',
            sizes: [{
                number: '36',
                available: false
            },
            {
                number: '37',
                available: false
            },
            {
                number: '38',
                available: true
            },
            {
                number: '39',
                available: true
            },
            {
                number: '40',
                available: true
            },
            {
                number: '41',
                available: true
            },
            {
                number: '42',
                available: false
            }
            ],
        },
        {
            id: 627,
            offer: "New",
            title: "Summer walk",
            sku: "AF18-36-Olive",
            stock: "in stock",
            price: 79,
            salePercentage: '',
            color: 'Olive',
            image: '../assets/shoes_1.jpg',
            sliderImages: [
                '../assets/shoes_1.jpg',
                '../assets/shoes_2.jpg',
                '../assets/shoes_3.jpg'
            ],
            sale: false,
            productRating: "5",
            selectedSize: '',
            sizes: [{
                number: '36',
                available: true
            },
            {
                number: '37',
                available: true
            },
            {
                number: '38',
                available: false
            },
            {
                number: '39',
                available: false
            },
            {
                number: '40',
                available: true
            },
            {
                number: '41',
                available: true
            },
            {
                number: '42',
                available: false
            }
            ]
        }
        ],
        gridImages: [{
            image: '../assets/women_shoes.jpg',
            header: "Women Collection",
            button: "Buy now"
        },
        {
            image: '../assets/men_shoes.jpg',
            header: "Men Collection",
            button: "Buy now"
        },
        {
            image: '../assets/children_shoes.jpg',
            header: "Children Collection",
            button: "Buy now"
        }
        ],
        sizeGuideMen: [{
            USA: [
                7, 7.5, 8, 8.5, 9, 9.5, 10, 10.5, 11, 11.5, 12, 13
            ],
            UK: [
                6, 6.5, 7, 7.5, 8, 8.5, 9, 9.5, 10, 10.5, 11, 12
            ],
            EUR: [
                39.5, 40, 41, 41.5, 42, 42.5, 43, 44, 44.5, 45, 46, 47
            ],
            cm: [
                26, 26.5, 27, 27.5, 28, 28.5, 29, 29.5, 30, 30.5, 31, 31.5
            ]
        }],
        sizeGuideWomen: [{
            USA: [
                6, 6.5, 7, 7.5, 8, 8.5, 9, 9.5, 10, 10.5, 11
            ],
            UK: [
                3.5, 4, 4.5, 5, 5.5, 6, 6.5, 7, 7.5, 8, 8.5
            ],
            EUR: [
                36.5, 37.5, 38, 38.5, 39, 39.5, 40, 41, 41.5, 42, 42.5
            ],
            cm: [
                24, 24.5, 25, 25.5, 26, 26.5, 27, 27.5, 28, 28.5, 29
            ]
        }],
        sizeGuideChildren: [{
            USA: [
                11, 12, 13, 1, 2, 3, 4, 5, 6
            ],
            UK: [
                10, 11, 12, 13, 1, 2, 3, 4, 5
            ],
            EUR: [
                28, 29.5, 31, 32, 33.5, 35, 36, 37.5, 38.5
            ],
            cm: [
                18.5, 19.5, 20, 21, 22, 23, 23.5, 24.5, 25.5
            ]
        }],
        inCart: [],
        productPreview: [],
        user: null
    },
    getters: {
        products: state => state.products,
        gridImages: state => state.gridImages,
        sizeGuideMen: state => state.sizeGuideMen,
        sizeGuideWomen: state => state.sizeGuideWomen,
        sizeGuideChildren: state => state.sizeGuideChildren,
        inCart: state => state.inCart,
        productPreview: state => state.productPreview,
        user: state => state.user
    },
    mutations: {
        CLEAR_PRODUCT_PREVIEW(state) { state.productPreview = [] },
        SHOW_PRODUCT(state, payload) { state.productPreview.push(payload) },
        SET_PRODUCT_SELECTED_SIZE(state, payload) { state.productPreview.map(product => product.selectedSize = payload) },
        ADD_TO_CART(state, payload) { state.inCart.push(payload) },
        REMOVE_FROM_CART(state, index) { state.inCart.splice(index, 1) },
        UPDATE_USER(state, { user }) { Vue.set(state, 'user', user) }
    },
    actions: {
        clearProductPreview(context) { context.commit('CLEAR_PRODUCT_PREVIEW') },
        showProduct(context, payload) { context.commit('SHOW_PRODUCT', payload) },
        setProductSelectedSize(context, payload) { context.commit('SET_PRODUCT_SELECTED_SIZE', payload) },
        addToCart(context, payload) { context.commit('ADD_TO_CART', payload) },
        removeFromCart(context, index) { context.commit('REMOVE_FROM_CART', index) }
    },
    plugins: [vuexLocalStorage.plugin]
});