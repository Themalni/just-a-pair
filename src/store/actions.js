import {
    SET_USER,
    CLEAR_USER,
    ADD_TO_CART,
    REMOVE_FROM_CART,
    SET_PRODUCT_TO_PREVIEW
} from "./types";

export default {
    [SET_USER]({commit}, { displayName, email }) {
        return commit(SET_USER, { displayName, email });
    },
    [CLEAR_USER]({commit}) {
        return commit(CLEAR_USER);
    },
    [ADD_TO_CART]({commit}, data) {
        return commit(ADD_TO_CART, data);
    },
    [REMOVE_FROM_CART]({commit}, index) {
        return commit(REMOVE_FROM_CART, index);
    },
    [SET_PRODUCT_TO_PREVIEW]({commit}, product) {
        return commit(SET_PRODUCT_TO_PREVIEW, product);
    }
}