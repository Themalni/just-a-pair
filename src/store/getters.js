export default {
    productPreview: state => state.productPreview,
    user: state => state.user
}