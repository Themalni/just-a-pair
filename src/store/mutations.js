import {
    SET_USER,
    CLEAR_USER,
    ADD_TO_CART,
    REMOVE_FROM_CART,
    SET_PRODUCT_TO_PREVIEW
} from "./types";

export default {
    [SET_USER](state, payload) {
        state.user = payload;
    },
    [CLEAR_USER](state) {
        state.user = null;
    },
    [ADD_TO_CART](state, payload) {
        const foundInCart = state.inCart.some(item => item.id === payload.id && item.size === payload.size);

        if(foundInCart) {
            state.inCart.map(item => {
                if(item.id === payload.id && item.size === payload.size) {
                    item.amount = item.amount + payload.amount;
                }
                return item;
            })
        } else {
            state.inCart.push(payload)
        }
    },
    [REMOVE_FROM_CART](state, {id, size}) {
        const productIndex = state.inCart.findIndex(cartItem => cartItem.id === id && cartItem.size === size);
        state.inCart.splice(productIndex, 1);
    },
    [SET_PRODUCT_TO_PREVIEW](state, payload) {
        state.productPreview = payload;
    }
}