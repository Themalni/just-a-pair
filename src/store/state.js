export default {
    selectedProduct: null,
    inCart: [],
    productPreview: null,
    user: null
}