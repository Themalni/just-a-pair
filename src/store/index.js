import Vue from 'vue'
import Vuex from 'vuex'
import VuexPersist from 'vuex-persist'
import state from './state'
import getters from './getters'
import mutations from './mutations'
import actions from './actions'

Vue.use(Vuex)

const vuexLocalStorage = new VuexPersist({ storage: localStorage })

export default new Vuex.Store({
  state,
  getters,
  mutations,
  actions,
  plugins: [vuexLocalStorage.plugin]
})
