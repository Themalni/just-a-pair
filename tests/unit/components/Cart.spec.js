import { mount, createLocalVue } from '@vue/test-utils';
import Cart from '@/components/Cart';
import mockState from '../utils/mockState';
import Vuex from 'vuex';
import salePrice from "@/filters.js";
import { REMOVE_FROM_CART } from "@/store/types";

const localVue = createLocalVue();
localVue.use(Vuex);
localVue.filter('salePrice', salePrice);

describe('Cart', () => {
    let wrapper
    let store
    let state

    beforeEach(() => {
        state = mockState.state
        store = new Vuex.Store({
            state: state
        })

        wrapper = mount(Cart, {
            store,
            localVue
        });
    });

    afterEach(() => {
        wrapper.destroy();
    });

    it('renders correctrly', () => {
        expect(wrapper.html()).toMatchSnapshot();
    });

    it('item is removed from cart', async () => {
        const removeFromCartMock = jest.spyOn(wrapper.vm, 'removeFromCart')
        const removeFromCartDispatchMock = jest.spyOn(wrapper.vm, REMOVE_FROM_CART);
        const button = wrapper.find('.btn-remove');

        await button.trigger("click");

        expect(removeFromCartMock).toHaveBeenCalled();
        expect(removeFromCartDispatchMock).toHaveBeenCalled();
    })

})