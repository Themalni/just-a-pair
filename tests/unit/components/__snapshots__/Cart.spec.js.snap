// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cart renders correctrly 1`] = `
<div>
  <div id="shoppingCartModal" tabindex="-1" role="dialog" aria-labelledby="exampleModalLabel" aria-hidden="false" class="modal fade">
    <div role="document" class="modal-dialog modal-frame modal-top">
      <div class="modal-content">
        <div class="modal-body">
          <table class="table">
            <thead class="main-blue white-text">
              <tr>
                <th scope="col" class="narrow-col">Image</th>
                <th scope="col" class="wide-col">Title</th>
                <th scope="col" class="narrow-col">Color</th>
                <th scope="col" class="narrow-col">Size</th>
                <th scope="col" class="narrow-col">Price</th>
                <th scope="col" class="narrow-col text-right">Amount</th>
                <th scope="col" class="narrow-col text-right">Item total</th>
                <th scope="col" class="narrow-col text-center">Remove</th>
              </tr>
            </thead>
            <tbody name="fade">
              <tr id="621" title="Summer sandals" sku="BR55-36-Black" price="120" salepercentage="30" color="Black" image="../assets/sneakers.jpg" sliderimages="../assets/sneakers.jpg,../assets/sneakers.jpg,../assets/sneakers.jpg" sale="true" amount="1" size="37" total="84">
                <th scope="row"><img src="../assets/sneakers.jpg" class="cart-item-img"></th>
                <td>Summer sandals</td>
                <td>Black</td>
                <td>37</td>
                <td>$84</td>
                <td class="text-right">1</td>
                <td class="text-right">$84</td>
                <td scope="col" class="text-center"><button class="btn light-blue btn-sm btn-remove">X</button></td>
              </tr>
            </tbody>
            <tfoot>
              <tr>
                <td colspan="6" class="text-right"><span class="mr-4">Total:</span> 1</td>
                <td colspan="1" class="text-right">$84</td>
                <td></td>
              </tr>
              <tr style="display: none;">
                <th colspan="5" class="text-center">Your shopping cart is emty</th>
              </tr>
            </tfoot>
          </table>
        </div>
        <div class="modal-footer"><button type="button" data-dismiss="modal" class="btn orange-accent btn-sm">Continue Shopping</button> <button type="button" class="btn main-blue btn-sm">Checkout</button></div>
      </div>
    </div>
  </div>
</div>
`;
